shared/src/androidMain/kotlin/dev/johnoreilly/confetti/work/SessionNotificationSender.kt

```diff
@@ -0,0 +1,162 @@
+package dev.johnoreilly.confetti.work
+
+import android.app.Notification
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.content.Context
+import android.graphics.BitmapFactory
+import android.os.Build
+import android.util.Log
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import com.apollographql.apollo3.cache.normalized.FetchPolicy
+import dev.johnoreilly.confetti.ConfettiRepository
+import dev.johnoreilly.confetti.auth.Authentication
+import dev.johnoreilly.confetti.fragment.SessionDetails
+import dev.johnoreilly.confetti.shared.R
+import dev.johnoreilly.confetti.utils.DateService
+import kotlinx.datetime.LocalDateTime
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.toInstant
+import kotlinx.datetime.toLocalDateTime
+import kotlin.time.Duration.Companion.minutes
+
+class SessionNotificationSender(
+    private val context: Context,
+    private val repository: ConfettiRepository,
+    private val dateService: DateService,
+    private val notificationManager: NotificationManagerCompat,
+    private val authentication: Authentication,
+) {
+
+    suspend fun sendNotification() {
+        Log.d("marcello", "notify")
+
+        // If there is no user signed-in, no reason to process all sessions.
+        val user = authentication.currentUser.value ?: return
+
+        val conferenceId = repository.getConference()
+        val conferenceDates = repository.conferenceData(conferenceId, FetchPolicy.CacheAndNetwork)
+            .data
+            ?.sessions
+            ?.nodes
+            ?.map { session -> session.sessionDetails.startsAt.date }
+            .orEmpty()
+
+        // If currente date is not in the conference range, no reason to process all sessions.
+        if (dateService.now().date !in conferenceDates) {
+            return
+        }
+
+        val bookmarks = repository.bookmarks(
+            conference = conferenceId,
+            uid = user.uid,
+            tokenProvider = user,
+            fetchPolicy = FetchPolicy.CacheAndNetwork,
+        )
+            .data
+            ?.bookmarks
+            ?.sessionIds
+            .orEmpty()
+
+        val sessions = repository.sessions(
+            conference = conferenceId,
+            uid = user.uid,
+            tokenProvider = user,
+            fetchPolicy = FetchPolicy.CacheAndNetwork,
+        )
+            .data
+            ?.sessions
+            ?.nodes
+            ?.map { it.sessionDetails }
+            .orEmpty()
+
+        val bookmarkedSessions = sessions.filter { session ->
+            bookmarks.contains(session.id)
+        }
+
+        val intervalRange = createIntervalRange()
+        val sessionsToNotify = bookmarkedSessions.filter { session ->
+            session.startsAt in intervalRange
+        }
+
+        if (sessionsToNotify.isNotEmpty()) {
```

Comment:
Since you have autoCancel, what's stopping this from recreating notifications the user has dismissed?

Reply:
There is nothing preventing recreation but that should not happen because:
1. The period work only runs each 15 minutes.
2. We only show the sessions "starting" in the next 15 minutes.

Not the most elegant solution. tho.

Reply:
I will double-check the logic but be sure I didn't miss anything.

Reply:
Hey @yschimke I checked and it is working as expected per message [here](https://github.com/joreilly/Confetti/pull/544#discussion_r1160686617).